File: backend/internal/domain/oauth2.go
package domain

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// ============================================================================
// OAuth2Provider - OAuth2 provider configuration (preset or custom)
// ============================================================================

// OAuth2Provider represents an OAuth2 provider configuration
type OAuth2Provider struct {
	ID   uuid.UUID `json:"id"`
	Slug string    `json:"slug"` // e.g., "google", "github", "slack"
	Name string    `json:"name"`

	// Endpoints
	AuthorizationURL string `json:"authorization_url"`
	TokenURL         string `json:"token_url"`
	RevokeURL        string `json:"revoke_url,omitempty"`
	UserinfoURL      string `json:"userinfo_url,omitempty"`

	// Configuration
	PKCERequired  bool     `json:"pkce_required"`
	DefaultScopes []string `json:"default_scopes"`

	// UI
	IconURL          string `json:"icon_url,omitempty"`
	DocumentationURL string `json:"documentation_url,omitempty"`

	// Metadata
	IsPreset  bool      `json:"is_preset"` // true = system-defined, false = custom
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// NewOAuth2Provider creates a new custom OAuth2 provider
func NewOAuth2Provider(slug, name, authURL, tokenURL string) *OAuth2Provider {
	now := time.Now().UTC()
	return &OAuth2Provider{
		ID:               uuid.New(),
		Slug:             slug,
		Name:             name,
		AuthorizationURL: authURL,
		TokenURL:         tokenURL,
		DefaultScopes:    []string{},
		IsPreset:         false,
		CreatedAt:        now,
		UpdatedAt:        now,
	}
}

// ============================================================================
// OAuth2App - Tenant-specific OAuth2 application configuration
// ============================================================================

// OAuth2App represents a tenant's OAuth2 application configuration
type OAuth2App struct {
	ID         uuid.UUID `json:"id"`
	TenantID   uuid.UUID `json:"tenant_id"`
	ProviderID uuid.UUID `json:"provider_id"`

	// Encrypted client credentials
	EncryptedClientID     []byte `json:"-"`
	EncryptedClientSecret []byte `json:"-"`
	ClientIDNonce         []byte `json:"-"`
	ClientSecretNonce     []byte `json:"-"`

	// Customization
	CustomScopes []string `json:"custom_scopes,omitempty"`
	RedirectURI  string   `json:"redirect_uri,omitempty"`

	// Status
	Status OAuth2AppStatus `json:"status"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// Relations (populated by repository)
	Provider *OAuth2Provider `json:"provider,omitempty"`
}

// OAuth2AppStatus represents the status of an OAuth2 app
type OAuth2AppStatus string

const (
	OAuth2AppStatusActive   OAuth2AppStatus = "active"
	OAuth2AppStatusDisabled OAuth2AppStatus = "disabled"
)

// NewOAuth2App creates a new OAuth2 app for a tenant
func NewOAuth2App(tenantID, providerID uuid.UUID) *OAuth2App {
	now := time.Now().UTC()
	return &OAuth2App{
		ID:         uuid.New(),
		TenantID:   tenantID,
		ProviderID: providerID,
		Status:     OAuth2AppStatusActive,
		CreatedAt:  now,
		UpdatedAt:  now,
	}
}

// GetScopes returns custom scopes if set, otherwise provider's default scopes
func (a *OAuth2App) GetScopes() []string {
	if len(a.CustomScopes) > 0 {
		return a.CustomScopes
	}
	if a.Provider != nil {
		return a.Provider.DefaultScopes
	}
	return []string{}
}

// ============================================================================
// OAuth2Connection - Individual OAuth2 token connection
// ============================================================================

// OAuth2Connection represents an OAuth2 token connection linked to a credential
type OAuth2Connection struct {
	ID           uuid.UUID `json:"id"`
	CredentialID uuid.UUID `json:"credential_id"`
	OAuth2AppID  uuid.UUID `json:"oauth2_app_id"`

	// Encrypted tokens
	EncryptedAccessToken  []byte `json:"-"`
	EncryptedRefreshToken []byte `json:"-"`
	AccessTokenNonce      []byte `json:"-"`
	RefreshTokenNonce     []byte `json:"-"`
	TokenType             string `json:"token_type"` // Usually "Bearer"

	// Expiration
	AccessTokenExpiresAt  *time.Time `json:"access_token_expires_at,omitempty"`
	RefreshTokenExpiresAt *time.Time `json:"refresh_token_expires_at,omitempty"`

	// OAuth2 flow state (temporary, used during authorization)
	State        string `json:"-"` // CSRF protection
	CodeVerifier string `json:"-"` // PKCE

	// Account information from userinfo endpoint
	AccountID    string          `json:"account_id,omitempty"`
	AccountEmail string          `json:"account_email,omitempty"`
	AccountName  string          `json:"account_name,omitempty"`
	RawUserinfo  json.RawMessage `json:"raw_userinfo,omitempty"`

	// Status
	Status        OAuth2ConnectionStatus `json:"status"`
	LastRefreshAt *time.Time             `json:"last_refresh_at,omitempty"`
	LastUsedAt    *time.Time             `json:"last_used_at,omitempty"`
	ErrorMessage  string                 `json:"error_message,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// OAuth2ConnectionStatus represents the status of an OAuth2 connection
type OAuth2ConnectionStatus string

const (
	OAuth2ConnectionStatusPending   OAuth2ConnectionStatus = "pending"
	OAuth2ConnectionStatusConnected OAuth2ConnectionStatus = "connected"
	OAuth2ConnectionStatusExpired   OAuth2ConnectionStatus = "expired"
	OAuth2ConnectionStatusRevoked   OAuth2ConnectionStatus = "revoked"
	OAuth2ConnectionStatusError     OAuth2ConnectionStatus = "error"
)

// NewOAuth2Connection creates a new OAuth2 connection
func NewOAuth2Connection(credentialID, oauth2AppID uuid.UUID) *OAuth2Connection {
	now := time.Now().UTC()
	return &OAuth2Connection{
		ID:           uuid.New(),
		CredentialID: credentialID,
		OAuth2AppID:  oauth2AppID,
		TokenType:    "Bearer",
		Status:       OAuth2ConnectionStatusPending,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
}

// IsAccessTokenExpired checks if the access token has expired (with 5 minute buffer)
func (c *OAuth2Connection) IsAccessTokenExpired() bool {
	if c.AccessTokenExpiresAt == nil {
		return false
	}
	// 5 minute buffer for proactive refresh
	return time.Now().Add(5 * time.Minute).After(*c.AccessTokenExpiresAt)
}

// IsRefreshTokenExpired checks if the refresh token has expired
func (c *OAuth2Connection) IsRefreshTokenExpired() bool {
	if c.RefreshTokenExpiresAt == nil {
		return false // No expiration = never expires
	}
	return time.Now().After(*c.RefreshTokenExpiresAt)
}

// CanRefresh checks if the connection can be refreshed
func (c *OAuth2Connection) CanRefresh() bool {
	return c.Status == OAuth2ConnectionStatusConnected &&
		c.EncryptedRefreshToken != nil &&
		!c.IsRefreshTokenExpired()
}

// MarkAsConnected updates the connection status to connected
func (c *OAuth2Connection) MarkAsConnected() {
	c.Status = OAuth2ConnectionStatusConnected
	c.ErrorMessage = ""
	c.UpdatedAt = time.Now().UTC()
}

// MarkAsExpired updates the connection status to expired
func (c *OAuth2Connection) MarkAsExpired() {
	c.Status = OAuth2ConnectionStatusExpired
	c.UpdatedAt = time.Now().UTC()
}

// MarkAsError updates the connection status to error
func (c *OAuth2Connection) MarkAsError(errMsg string) {
	c.Status = OAuth2ConnectionStatusError
	c.ErrorMessage = errMsg
	c.UpdatedAt = time.Now().UTC()
}

// MarkAsRevoked updates the connection status to revoked
func (c *OAuth2Connection) MarkAsRevoked() {
	c.Status = OAuth2ConnectionStatusRevoked
	c.UpdatedAt = time.Now().UTC()
}

// UpdateTokens updates the access and refresh tokens after a refresh
func (c *OAuth2Connection) UpdateTokens(encryptedAccessToken, encryptedRefreshToken, accessNonce, refreshNonce []byte, expiresAt *time.Time) {
	now := time.Now().UTC()
	c.EncryptedAccessToken = encryptedAccessToken
	c.AccessTokenNonce = accessNonce
	c.AccessTokenExpiresAt = expiresAt
	c.LastRefreshAt = &now

	if encryptedRefreshToken != nil {
		c.EncryptedRefreshToken = encryptedRefreshToken
		c.RefreshTokenNonce = refreshNonce
	}

	c.Status = OAuth2ConnectionStatusConnected
	c.ErrorMessage = ""
	c.UpdatedAt = now
}

// RecordUsage records when the token was last used
func (c *OAuth2Connection) RecordUsage() {
	now := time.Now().UTC()
	c.LastUsedAt = &now
	c.UpdatedAt = now
}

// ============================================================================
// OAuth2 Authorization Request/Response
// ============================================================================

// OAuth2AuthorizationRequest represents an authorization request
type OAuth2AuthorizationRequest struct {
	ProviderSlug string      `json:"provider_slug"`
	Scope        OwnerScope  `json:"scope"`
	ProjectID    *uuid.UUID  `json:"project_id,omitempty"`
	Name         string      `json:"name"`
	Scopes       []string    `json:"scopes,omitempty"` // Additional scopes
}

// OAuth2AuthorizationResponse represents the response for starting authorization
type OAuth2AuthorizationResponse struct {
	AuthorizationURL string `json:"authorization_url"`
	State            string `json:"state"`
}

// OAuth2TokenResponse represents the response from token endpoint
type OAuth2TokenResponse struct {
	AccessToken  string `json:"access_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int    `json:"expires_in,omitempty"`
	RefreshToken string `json:"refresh_token,omitempty"`
	Scope        string `json:"scope,omitempty"`

	// Error fields
	Error            string `json:"error,omitempty"`
	ErrorDescription string `json:"error_description,omitempty"`
}

// OAuth2UserinfoResponse represents common fields from userinfo endpoint
type OAuth2UserinfoResponse struct {
	Sub           string `json:"sub,omitempty"`
	Email         string `json:"email,omitempty"`
	EmailVerified bool   `json:"email_verified,omitempty"`
	Name          string `json:"name,omitempty"`
	Picture       string `json:"picture,omitempty"`

	// GitHub specific
	Login string `json:"login,omitempty"`
	ID    int64  `json:"id,omitempty"`

	// Full response for storage
	Raw json.RawMessage `json:"-"`
}

// GetAccountID returns the account ID from various providers
func (u *OAuth2UserinfoResponse) GetAccountID() string {
	if u.Sub != "" {
		return u.Sub
	}
	if u.Login != "" {
		return u.Login
	}
	if u.ID != 0 {
		return string(rune(u.ID))
	}
	return ""
}

// GetDisplayName returns the display name from various providers
func (u *OAuth2UserinfoResponse) GetDisplayName() string {
	if u.Name != "" {
		return u.Name
	}
	if u.Login != "" {
		return u.Login
	}
	if u.Email != "" {
		return u.Email
	}
	return ""
}

File: frontend/types/oauth2.ts
/**
 * OAuth2 Types
 * 外部サービス認証機能の型定義
 */

// OAuth2 Provider (Preset or custom OAuth2 provider)
export interface OAuth2Provider {
  id: string
  slug: string
  name: string
  icon_url?: string
  authorization_url: string
  token_url: string
  revoke_url?: string
  userinfo_url?: string
  pkce_required: boolean
  default_scopes: string[]
  available_scopes?: string[]
  description?: string
  authorization_type?: string
  documentation_url?: string
  is_preset: boolean
  created_at: string
  updated_at: string
}

// OAuth2 App (Tenant's OAuth2 client configuration)
export interface OAuth2App {
  id: string
  tenant_id: string
  provider_id: string
  provider_slug: string
  provider?: OAuth2Provider
  client_id: string
  scopes?: string[]
  custom_scopes?: string[]
  redirect_uri?: string
  status: 'active' | 'disabled'
  created_at: string
  updated_at: string
}

// OAuth2 Connection Status
export type OAuth2ConnectionStatus = 'pending' | 'connected' | 'expired' | 'revoked' | 'error'

// OAuth2 Connection (Individual token/connection)
export interface OAuth2Connection {
  id: string
  credential_id: string
  oauth2_app_id: string
  token_type: string
  access_token_expires_at?: string
  refresh_token_expires_at?: string
  account_id?: string
  account_email?: string
  account_name?: string
  status: OAuth2ConnectionStatus
  last_refresh_at?: string
  last_used_at?: string
  error_message?: string
  created_at: string
  updated_at: string
}

// Credential Scope (where credentials can be used)
export type CredentialScope = 'organization' | 'project' | 'personal'

// Share Permission (level of access when sharing)
export type SharePermission = 'use' | 'edit' | 'admin'

// Credential Share (sharing a credential with user or project)
export interface CredentialShare {
  id: string
  credential_id: string
  target_user_id?: string
  target_user_email?: string
  target_project_id?: string
  target_project_name?: string
  shared_with_user_id?: string
  shared_with_project_id?: string
  permission: SharePermission
  shared_by_user_id: string
  note?: string
  expires_at?: string
  created_at: string
}

// API Request Types

export interface CreateOAuth2AppRequest {
  provider_slug: string
  client_id: string
  client_secret: string
  scopes?: string[]
  custom_scopes?: string[]
  custom_authorization_url?: string
  custom_token_url?: string
}

export interface StartAuthorizationRequest {
  provider_slug: string
  name: string
  scope: CredentialScope
  project_id?: string
  scopes?: string[]
}

export interface StartAuthorizationResponse {
  authorization_url: string
  state: string
  credential_id: string
}

export interface ShareWithUserRequest {
  target_user_email: string
  shared_with_user_id?: string
  permission: SharePermission
  note?: string
  expires_at?: string
}

export interface ShareWithProjectRequest {
  target_project_id: string
  shared_with_project_id?: string
  permission: SharePermission
  note?: string
  expires_at?: string
}

export interface UpdateShareRequest {
  permission?: SharePermission
  note?: string
  expires_at?: string
}

// Provider Response (with app configuration status)
export interface OAuth2ProviderWithStatus extends OAuth2Provider {
  app_configured: boolean
}

// Required Credential (from BlockDefinition)
export interface RequiredCredential {
  name: string
  type: 'api_key' | 'oauth2' | 'bearer' | 'basic' | 'custom'
  scope: 'system' | 'tenant'
  description: string
  required: boolean
}

// Credential Binding (mapping in Step)
export interface CredentialBinding {
  name: string
  credential_id: string
}
